lambda/api_update_av.tf

```diff
@@ -0,0 +1,75 @@
+resource "aws_lambda_function" "lambda_api_update_function" {
```

Comment:
This seems really tied to the TDR API implementation. My understanding of this repo was that it was for code that could be shared by multiple Terraform projects.

Does it make sense for the implementation-specific code to be in tdr-terraform-environments, and keep this repo very generic so that we can use it to create buckets, lambdas, etc. with secure defaults?

I'd be interested in @paulschwarzenberger's thoughts on this when he's around.

Reply:
Yeah, we can ask him tomorrow as this was his idea, to have a specific tf file for each lambda.

Reply:
We had a discussion about this issue this morning:

- It's hard to split the lambda modules into TDR-specific code and generic module code like we could for the S3 bucket.
- I personally think this makes it harder for us to use the modules in other projects, but we think it's worth the trade-off of not having to split out the generic lambda code and the TDR-specific code
- Having some TDR-specific code in this repo isn't a bad thing, because people adding new code (e.g. a new lambda) can copy an existing well-thought-through example rather than writing it from scratch.
- If we'd had the modules from the start, we might have put _more_ TDR-specific code into here rather than tdr-terraform-environments. (Though I'm afraid I'm still not clear about how we'd decide what goes into this repo and what goes into the environments one.)

Paul and Sam, is that a fair summary?

I'd still wonder whether it would be possible to have multiple lambda modules rather than a single module with all the boolean parameters in the lambda function - if this were code, it would be like a function that does six different things depending on the parameter, and you'd split it into multiple functions. What I'd really like to do is to have a go at refactoring it to see if I can make a concrete suggestion, but realistically I don't have time to look into it now. 😅

Reply:
Yes, good summary. At the moment, the repository is structured with a single folder per AWS service, so it's really clear for a developer to see which module to use for a resource, and encourages a uniform approach. Also, splitting the lambda module into multiple modules, one for each lambda, could mean duplication of common items such as the IAM policy templates.